command/base.go

```diff
@@ -333,6 +333,9 @@ func (c *BaseCommand) flagSet(bit FlagSetBit) *FlagSets {
 				addrStringVar.Default = c.flagAddress
 			} else {
 				addrStringVar.Default = "https://127.0.0.1:8200"
+				if os.Getenv("VAULT_ADDR") == "" {
```

Comment:
Isn't this going to be used in Agent as well, where it's normal/desirable to have VAULT_ADDR unset?

Reply:
True - it may I didnt think of that actually - although I did try it in the case of server where same thing happens and does seem good and a nice warning the same where it defaults to binding to loopback. I will test / try it and confirm.

@ncabatoff  But in any case what would be an exclusionary condition there to discard the evaluation in the case of agent?

Reply:
Whoops, silly me to forget about vault server.  Why do you think it's desirable there?

> what would be an exclusionary condition there to discard the evaluation in the case of agent?

You're asking "how do I make this not apply to agent and server"?  Not sure what the best approach would be.  One idea: add a bool field to BaseCommand, e.g. `suppressMissingAddrWarning`, governing this behaviour.  Have server and agent create BaseCommand with that field set to true.

Reply:
> Why do you think it's desirable there?

In the case of Vault server - In my highly opinioned approach - wherever possible bind address should be detailed in the (HCL) listener stanza or as a last resort in the case of `-dev` inlined with related parameter: `-dev-listen-addres` or otherwise in regular server mode with `-address`.

Regardless of this PR - at present there's sufficient details in the boot headers showing the current listen / bind address of server (when Vault starts). But leaving this message in would add an extra line of warning with consistency the same as Vault clients when VAULT_ADDR is expected to be set (or configured) but is not.

In the case of `server` this warning would only be applicable when launched without either VAULT_ADDR or `-address` being set and where it defaults to loopback.

> ... add a bool field to BaseCommand, e.g. `suppressMissingAddrWarning`,

This proposal seems more complete - let me see if I can update this PR as per these intentions.

Reply:
I don't think server/agent should be consistent with the rest of the CLI regarding VAULT_ADDR.

Reply:
> I don't think server/agent should be consistent with the rest of the CLI regarding VAULT_ADDR.

I could agree with that - but on the other hand where does it detail this expected fall back behaviour to use loopback / 127.0.0.1? - I do not see this explicitly detailed anywhere in our docs or elsewhere - so in that sense a warning detailing why this has happened would make it more evident and also consistent with CLI clients.

I may be wrong and this may be already detailed somewhere on some docs but nonetheless I think explicitly detailing this expected behaviour as part of the startup operational logs would be better than it being assumed or known (because it's already on some other external docs).

Reply:
>  ... could you elaborate please?

Where you do not specify any `-address`, `listener` on launching server / agent - it defaults to `127.0.0.1` which is then the API / RPC address (falls back to that). However at present it's not clear if this is implicitly caused by no configuration values being passed or if `127.0.0.1` is what's actually set in either `-address` or `listener`.

In any case I think we can strive for what the community has requested and what can generally help practitioners with CLI who often forget to initial set or change addresses appropriately.

Further to the earlier suggestion with `BaseCommand` - I've had a look at commands/* function and it would seem that the same parameter would need to be set per command like `status`, `server`, `agent`, etc. 

Also since everything is calling [flagSet](https://github.com/hashicorp/vault/blob/main/command/base.go#L302) - I wasn't able to tie it all together as [adding `flagMuteAddrWarn` to BaseCommand](https://github.com/hashicorp/vault/blob/main/command/base.go#L36-L71) and adjusting it per command (`server`) only gets updated or set later.

Hey @ncabatoff - I have another working approach that [uses the `runtime` package](https://pkg.go.dev/runtime) which I'd like to check with you before just committing. Those [additions to `flagSet(`](https://github.com/hashicorp/vault/blob/main/command/base.go#L302) would be:

```
import (
// ...
	"runtime"
// ...
// flagSet creates the flags for this command. The result is cached on the
// command to save performance on future calls.
func (c *BaseCommand) flagSet(bit FlagSetBit) *FlagSets {
	_, fileCaller, _, _ := runtime.Caller(1)
// ... 
// ... truncated
// ... 
			if c.flagAddress != "" {
				addrStringVar.Default = c.flagAddress
			} else {
				addrStringVar.Default = "https://127.0.0.1:8200"
				// Show Address Warning except for server or agent
				if !strings.HasSuffix(fileCaller, "server.go") && !strings.HasSuffix(fileCaller, "agent.go") {
					if os.Getenv("VAULT_ADDR") == "" {
						c.UI.Warn(wrapAtLength(fmt.Sprintf("WARNING! VAULT_ADDR and -address unset. Defaulting to %s.", addrStringVar.Default)))
					}
				}
			}

```

Reply:
Re runtime: I tend to avoid clever solutions like this with implicit behaviour.  I'm also not fond of tying it to specific filenames; if we were to take this path I'd rather that we look at package name.  If we had no other options I'd be ok with this approach, but I'm unconvinced my original suggestion won't work.  Let's chat.

Reply:
Resolved to now show when `agent` or `server`.

I was not correctly initialising `MuteAddrWarning` (that's not a flag rather initialised in `commands.go`). Many thanks to @ncabatoff for clarifying this.